FreeBSD port, new features

ke6jjj opened the pull request:
Hey Dan, I sent you a message about these new changes. Then I just now saw the "push request" button. So I pushed it. See the e-mail for more details. But, in particular I feel I made some rude changes to the Makefile that you should feel free to change back (I wanted other CFLAGS and a different compiler). I just forgot about them when I committed them back to github.

-J

Commit:
Start of port to FreeBSD.

Commit:
Ported to FreeBSD (as far as I can tell).

Commit:
Fixed a few misunderstandings:
- Internal route messages and interface address ioctl() calls
  must be made using network ordering. Changed to make sure all
  calls to kernel do so. (All internal addresses in this program
  are in host order).
- The same goes for the "localip" configured with what used to be
  the "-L" flag.
Removed the '-L' flag and made local ip address a required argument.
(There's no point having a default because it is very likely wrong
for everyone else).

Commit:
- Changed tunnel creation so that each tunnel's inner addresses are
  separate from its outer addresses. The outer addresses retain the
  classical use (both being publically routable non-AMPR addresses).

  The inner addresses of a tunnel are now intended to be AMPR-only
  addresses. The local inner address now comes from the command line
  via a new, required argument required. The remote inner address is
  now synthesized from the first network advertised on that tunnel.
  (There might be more work to do with this, particularly if an
  update is received and the routed network is no longer associated
  with the tunnel).

  This change cleans up many routing loops and other problems that
  came from using the same address on the inside of each tunnel as
  on the outside. It may be that no separate routing tables are needed
  when running the daemon. (I no longer use separate tables for the
  tunnels and routes).

- Added a new required argument: local tunnel ip.

- Network routes are now directed straight to the tunnel interface
  hosting that network rather than the older method of specifying the
  outer tunnel address as the gateway. This also reduces routing
  loops and other confusions.

- Completed a feature which allows the administrator to choose which
  routes will be accepted and which will be ignored. All route
  advertisements for networks contained within a CIDR block given via
  the -A option will be accepted (and consequently, tunnels for those
  networks will be made).

  All route advertisements for networks contain within a CIDR block
  given via the -I option will be ignored.

  When acceptance and ignorance blocks overlap, the most specific
  specification will win.

  If no acceptance or rejection specifications are given, the
  default will be to accept all routes.

  A handy example might be: -A 44.0.0.0/8 -I 44.4.39.8/29
  This would generally accept all AMPR routes and reject all others.
  However, if a route is advertised within the network 4.4.39.8, which
  is my local allocation, it will be ignored.

  These option is now the preferred way to ignore local routes
  (as opposed to ignoring them based on the advertised tunnel
  address).

- There are now two route table options. The first is the classical
  route table, which controls the route table that tunnel interfaces
  use and the table into which the routes announced by each tunnel
  entered. (Specified by the -T option).

  The second, newer, option controls the route table used when creating
  the RIPv2 route reception socket. (Specified via the -B option). This
  second option is helpful in forcing the route reception socket to
  get the right interface to become a member of the RIPv2 multicast
  group on start up. (The "ifmcstat" utility can be used to check if
  this is so.)

- Fixed paste-o in message logging which caused most messages to
  be logged as LOG_INFO. Now errors and fatal errors are logged as
  LOG_ERR.

- Added testing option. If -f is specified, the given file will
  be treated as a RIPv2 packet and processed as if it had been
  received over the network. (May remove this or clean it up).

- Code cleanup: made all global variables static, removed an unused
  global (lowgif).

Commit:
Added a startup script.

Commit:
Added self to copyright holders.
Reduced logging severity of network skipping from notice to info.

Commit:
Added start for man page.

Commit:
- Fixed bug in netsted gateway detection (gateway inside the network
  it advertises).
- Changed calls to err() into calls to fatal() so that error
  messages make it into syslog.
- Added "fatale()", which mimicks err() by appending errno
  message.
- Added some debugging statements for logging why chroute() is
  failing.

Commit:
- Added ability to resume state by examing the current system for
  existing tunnels and routes.
- Reorganized code into separate modules a bit better and created
  separate header files for each module.
- Fixed bug in bitset/bitclr where new allocations were not getting
  cleared. (reallocarray() does not clear new memory).
- Added a return type to the visitor function in ipmapdo(); if
  the visitor returns true then all further iteration stops and
  ipmapdo() returns to the caller. (Can save time if the caller was
  just looking for a specific condtion).

There is currently still a bug with this code in that it can't seem
to find any acceptable routes. The filtering logic for route searching
just needs to be tweaked. (No netmasks appear to be acceptable).

Commit:
Fixed final problems parsing existing routes. 1) Sockt address
structures in route messages are apparently aligned to 'long'
boundaries. 2) If the sa_len of a netmask address is zero, it means
that the entire mask is zero. (Weird shorthand).
Seems to boot up ok! Going to try it.

Commit:
Fixed bug where discovered routes would expire immediately on
reception of next packet. Discovered routes are now assumed to have
been added when the daemon starts up and discovers them, and thus,
receive expiration dates in the future (rather than on January 1, 1970).

Commit:
- Changed discovery phase to be a single function. The OS needs the
  two sub-phases to be under its control, so it makes no sense for
  the OS-independent code to be in charge of calling them one after
  another.
- (OS) Added a temporary table to keep track of tunnels seen during the
  interface discovery so that routes received without interface names
  but with interface index IDs can still be understood.
- (OS) Split up some large functions into smaller pieces to make the
  discovery process easier to read.
- (OS) Cleaned up route message parsing code to make it easier to
  read. Added extensive comments. (Perhaps too many).
- Renamed "fatale()" to "fatal_err()" so it sounds less like an
  adjective.
- (OS) Fixed bug where local AMPR addresses appeared to be handled
  by tunneling interfaces when in fact they were actually bound
  to the loopback interface. Solution is to examine route messages
  deeper than just the "RTA_GATEWAY" address by looking for an
  RTA_IFP address if provided.
- (OS) Fixed bug where discovered routes to hosts were interpreted as
  routes to everywhere (0 netmask).
- (OS) Added more information to fatal error messages when having
  trouble establishing and removing routes.
- (Lib) Fixed bug in ipmapnearest() which caused it to return data
  from internal branch nodes when it should instead return the last
  node it found with valid data.
- (Lib) Added ipmapdotopdown(), which works like ipmapdo() except that
  it iterates from least-specific (smaller key lengths) to
  most-specific entries.
- Added a cleanup phase to remove redundant (overlapping) routes after
  discovery. Uses new ipmapdotopdown() function.
- Added -D flag to dump the current configuration (tunnels, routes and
  accept/reject policy). Useful for determining what the program thinks
  the current routing situation is.
- Fixed broken test case builds by making sure the new split modules
  are built with them and by making sure the tests use the new header
  files.
- Changed OS-independent discovery callbacks to use a context structure
  when updating the system with discovered items, in a small attempt to
  rely less on global variables where possible.

Commit:
(Forgot) Forgot to commit Makefile previously.
- Fixed bug in route change detection that I introduced earlier.
  There still remains the issue of coalescing a more-specific route
  back into a tunnel with a less-specific route that covers it.
  I'm sure this happens from time to time and it should be easy to
  discover and implement.

Commit:
Added logic to safely move and remove routes given the complicated
  way that the BSD kernel requires us to have remote inner addresses
  on tunnels (even though they are never used) and explicitly ties
  down the routes to said addresses. It also completely tears up any
  routes made to an interface whenever that interface's inner addresses
  change, even if the route was specifically targeted to the interface
  and not necessarily its inner addresses.
- Added logic to tear down any useless tunnels at startup.
- Moved a call to rmroute() before the call to unlinkroute()
  in destroy() in order to preserve the invariant that a tunnel
  is the first to know about routes being added or removed from it.

dancrossnyc commented on lib.c:
```diff
@@ -72,9 +61,14 @@ ipmapnearest(IPMap *map, uint32_t key, size_t keylen)
-			return map->datum;
-		parent = map;
+		if (keylen == 0) {
+			if (map->datum != NULL)
```
The intent here is to skip over empty interior nodes?

ke6jjj replied:
Yes.

ke6jjj committed:
- Fixed a bothersome bug in linkroute() which was screwing up the
  detection of existing routes, which then caused them to be 'added'
  when they really should have been 'changed'. (linkroute() was
  setting the route's gateway to the inner tunnel address, but routes
  in this system are stored and searched by their outer tunnel address
  because this is the 'nexthop' address given in the RIP updates).
- Fixed tunnel rebasing (adopting a new inner address when the route
  that was used to establish the tunnel is moving to a new tunnel) so
  that the interface's inner addresses are removed completely rather
  than having the tunnel interface being marked as 'down'. Bringing
  an interface down does not make it loose its routing entries.
  Changing its inner address, however, does.
- Changed instances where cidr block sizes were being stored in
  size_t types to being stored in int types. netmask2cidr(), which
  is the primary function where these values are generated, returns
  an int and not a size_t. Errors in cidr interpretation return -1,
  which is not detectable in an accurate and portable manner
  in a size_t. (This caused a few logging statements to change from
  using %zu to %d).
- Adopted Dan's coding style for if statements with complicated
  multi-line expressions by having the statement block's curly
  braces start on a line of their own.
There are probably some excessive debugging statements that can
be removed from the code, should the route add/move/delete problems
be finally fixed.

ke6jjj committed:
Fixed bug which prevented tunnels from receiving inner address assignments.

ke6jjj committed:
Fixed some uint32_t vs. size_t disagreements that slipped in and are only
caught when compiling on a 64-bit platform.

ke6jjj committed:
Finally fixed long-standing annoying bug triggered by route change
detection.  Bug was triggered when:

1. A previously routed network is detected as having been assigned to a new
   tunnel, and
2. The losing tunnel's inner address comes entirely from the routed network it
   it is losing.

This particular change requires a "rebase" operation, in which the losing
tunnel is brought down, brought back up, and then reassigned all of the
networks that it previously had, minus the network that is migrating away from
it.

However, due to a typo, while the rebasing logic thought it was adding all the
old undisturbed networks back to the tunnel, it in fact was adding the very
network it was supposed to loose in the first place, and not just once, but
over and over again.

(Use of wrong variable, "route" when it should have used "other").

ke6jjj committed:
Changed startup script to be more compatible with "service" utility.
Also fixed startup message so that it stays entirely on one line.

ke6jjj committed:
Fixed a bug in route change logic when a network is being moved to a new tunnel
which was specifcially brought up just for that network. In this case is is
important not to attempt to programmatically add the network to the tunnel
because the kernel will already have done so when the tunnel was brought up.
Similar logic for this condition already exists in addroute(), it just needed
to be added to chroute() as well.
